components/script/dom/resizeobserver.rs

```diff
@@ -0,0 +1,260 @@
+            .borrow_mut()
+            .retain_mut(|(_obs, other)| !(&**other == target));
+    }
+
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobserver-disconnect>
+    fn Disconnect(&self) {
+        self.observation_targets.borrow_mut().clear();
+    }
+}
+
+/// State machine equivalent of active and skipped observations.
+#[derive(Default, JSTraceable, MallocSizeOf)]
+enum ObservationState {
+    #[default]
+    Done,
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobserver-activetargets-slot>
+    Active,
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobserver-skippedtargets-slot>
+    Skipped,
+}
+
+/// https://drafts.csswg.org/resize-observer/#resizeobservation
+#[derive(JSTraceable, MallocSizeOf)]
+struct ResizeObservation {
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobservation-target>
+    /// Note: `target` is kept out of here, to avoid having to root the `ResizeObservation`.
+
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobservation-observedbox>
+    observed_box: ResizeObserverBoxOptions,
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobservation-lastreportedsizes>
+    last_reported_sizes: DomRefCell<VecDeque<ResizeObserverSizeImpl>>,
+    /// State machine mimicking the "active" and "skipped" targets slots of the observer.
+    state: ObservationState,
+}
+
+impl ResizeObservation {
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobservation-resizeobservation>
+    pub fn new(observed_box: ResizeObserverBoxOptions) -> ResizeObservation {
+        ResizeObservation {
+            observed_box,
+            last_reported_sizes: Default::default(),
+            state: Default::default(),
+        }
+    }
+
+    /// <https://drafts.csswg.org/resize-observer/#dom-resizeobservation-isactive>
+    fn is_active(&self, target: &Element) -> bool {
+        let reported_sizes = self.last_reported_sizes.borrow();
+        let Some(size) = reported_sizes.back() else {return true};
+        let box_size = calculate_box_size(target, &self.observed_box);
+        let width = box_size.width().to_f64_px();
+        let height = box_size.height().to_f64_px();
+        !((size.inline_size(), size.block_size()) == (width, height))
+    }
+}
+
+/// <https://drafts.csswg.org/resize-observer/#calculate-depth-for-node>
+fn calculate_depth_for_node(target: &Element) -> ResizeObservationDepth {
+    let node = target.upcast::<Node>();
+    let depth = node.ancestors().count();
+    ResizeObservationDepth(depth)
+}
+
+/// <https://drafts.csswg.org/resize-observer/#calculate-box-size>
+fn calculate_box_size(target: &Element, observed_box: &ResizeObserverBoxOptions) -> Rect<Au> {
+    // TODO: batch into one layout query for list of elems?
+    match observed_box {
+        ResizeObserverBoxOptions::Content_box |
+        ResizeObserverBoxOptions::Device_pixel_content_box => {
+            target.upcast::<Node>().bounding_content_box_or_zero()
```

Comment:
Device_pixel_content_box should snap to device pixels.
Also, `bounding_content_box_or_zero` is the wrong choice if there are multiple fragments, we don't want the bounding rect. As per the current spec it should just be the 1st fragment, with the idea of extending to an array with all fragments in the future.

Reply:
Thanks! I've added your comment to https://github.com/servo/servo/issues/31182 because I think some of these might best be addressed as follow-ups, so that this PR can focus more on the overal mechanism and get the event-loop integration right. 

Would you happen to have links to the code where one could get the right layout queries or transform the resulting data into the right format?

Reply:
It should be possible to use the new query API that's been exposed to implement this correct versus in a followup

Reply:
> use the new query API

Do you mean `content_boxes`? If so, I have added a commit that uses that one.